### ISSUE TITLE
Jenkins slave never successfully connects to the Jenkins master

### ISSUE THREAD

**Issue by barkerd427** [jenkinsci/mesos-plugin#83]
At least once a day, a Jenkins slave never actually registers with the master. It has downloaded the appropriate jar, but it is then unable to connect to the master. There should be some method of cleaning up orphaned slaves that are unable to connect. I'm not sure if it's as simple as updating the slave jar to kill itself after a certain amount of time given the right argument, or if it is better to include some management system in this plugin.

**Comment by barkerd427** [comment]
Apparently sometimes it actually does connect, but then never gets removed or never appears on the master. It is rather unpredictable behavior, and I can't reliably recreate it.

**Comment by vinodkone** [comment]
What do the jenkins slave logs say about why it's unable to connect?

**Comment by barkerd427** [comment]
Sometimes I get this:
```
WARNING: Logging before InitGoogleLogging() is written to STDERR
I0130 12:12:15.611502 20088 fetcher.cpp:76] Fetching URI 'http://build.aeon.cerner.corp/jnlpJars/slave.jar'
I0130 12:12:15.611665 20088 fetcher.cpp:126] Downloading 'http://build.aeon.cerner.corp/jnlpJars/slave.jar' to '/tmp/mesos/slaves/20150113-160441-1802747402-5050-1104-S23/frameworks/20150113-160441-1802747402-5050-1104-0000/executors/mesos-jenkins-d62b899a-cfc8-4ab2-a69b-409faeb2af3c/runs/f6711b8c-791e-42da-9554-0b8788547e51/slave.jar'
I0130 12:12:15.692030 20088 fetcher.cpp:255] Skipped extracting path '/tmp/mesos/slaves/20150113-160441-1802747402-5050-1104-S23/frameworks/20150113-160441-1802747402-5050-1104-0000/executors/mesos-jenkins-d62b899a-cfc8-4ab2-a69b-409faeb2af3c/runs/f6711b8c-791e-42da-9554-0b8788547e51/slave.jar'
I0130 12:12:16.835422 20140 exec.cpp:132] Version: 0.21.1
I0130 12:12:16.842278 20149 exec.cpp:206] Executor registered on slave 20150113-160441-1802747402-5050-1104-S23
Jan 30, 2015 12:12:17 PM hudson.remoting.jnlp.Main createEngine
INFO: Setting up slave: mesos-jenkins-d62b899a-cfc8-4ab2-a69b-409faeb2af3c
Jan 30, 2015 12:12:17 PM hudson.remoting.jnlp.Main$CuiListener <init>
INFO: Jenkins agent is running in headless mode.
Jan 30, 2015 12:12:17 PM hudson.remoting.jnlp.Main$CuiListener status
INFO: Locating server among [http://build.aeon.cerner.corp/]
Jan 30, 2015 12:12:17 PM hudson.remoting.jnlp.Main$CuiListener status
INFO: Connecting to build.aeon.cerner.corp:47151
Jan 30, 2015 12:12:17 PM hudson.remoting.jnlp.Main$CuiListener status
INFO: Handshaking
Jan 30, 2015 12:12:17 PM hudson.remoting.jnlp.Main$CuiListener status
INFO: Connected
Jan 30, 2015 12:24:12 PM hudson.remoting.jnlp.Main$CuiListener status
INFO: Terminated
Jan 30, 2015 12:24:12 PM jenkins.slaves.restarter.JnlpSlaveRestarterInstaller$2$1 onDisconnect
INFO: Restarting slave via jenkins.slaves.restarter.UnixSlaveRestarter@434f28de
Failing to obtain http://build.aeon.cerner.corp/computer/mesos-jenkins-d62b899a-cfc8-4ab2-a69b-409faeb2af3c/slave-agent.jnlp
java.io.IOException: Failed to load http://build.aeon.cerner.corp/computer/mesos-jenkins-d62b899a-cfc8-4ab2-a69b-409faeb2af3c/slave-agent.jnlp: 404 Not Found
	at hudson.remoting.Launcher.parseJnlpArguments(Launcher.java:275)
	at hudson.remoting.Launcher.run(Launcher.java:219)
	at hudson.remoting.Launcher.main(Launcher.java:192)
Waiting 10 seconds before retry
```

So it looks like it is connected at some point, and other times it never gets the connection. It's as if the slave on Jenkins is already deleted before Mesos can start the slave and the slave can communicate back. I will note that the slaves that output the above do generally die, but those that never achieve a connection live until I kill them.

**Comment by vinodkone** [comment]
Interesting. Is "UnixSlaveRestarter" a plugin that you use? Anyway, it would be interesting to correlate this slave related log lines on jenkins master log to see why jenkins master decided to terminate it. That might give us a clue into whats happening.

**Comment by maselvaraj** [comment]
@barkerd427 Have you tried '-noReconnect' option in 'Additional Jenkins Slave Agent JNLP arguments' ?

**Comment by barkerd427** [comment]
I'll give that a try. I haven't really found anything in the logs to
indicate why these errors are occurring, but this could be a solution.

**Comment by maselvaraj** [comment]
I am also not sure why the slave got disconnected. But '-noReconnect' might help you so that the slave won't reconnect automatically which might overcome the problem of manually killing the slave process to reclaim resources.
Please let me know how it goes.

**Comment by barkerd427** [comment]
This seems to be working, but I will wait to close this issue until the end of the week to ensure it is well tested under load.

**Comment by maselvaraj** [comment]
sure. thanks!

**Comment by barkerd427** [comment]
I am going to close this as -noReconnect solving the problem. Those slaves that didn't have that argument continued to have problems while the slaves with that argument never experienced any issues.

**Comment by maselvaraj** [comment]
@barkerd427 By any chance was your Jenkins hogging the resource offers sent by Mesos when you had this issue? I had a strange issue in my cluster where one Jenkins swallowing all the resources leaving others starving. Looking into the logs, in one of the tasks I saw 
"INFO: Restarting slave via jenkins.slaves.restarter.UnixSlaveRestarter..."

**Comment by barkerd427** [comment]
@maselvaraj, I did experience that. When the slave was killed by Mesos, Jenkins would randomly be able to restart it before the container died. I never figured out why it was random, but I assumed there was some network slowness that allowed most to die before they could be reconnected. If they "reconnected" to the master, then they remained running and registered on Mesos. The reason reconnected is in quotes is because they were not actually showing as connected on Jenkins. I'm assuming this has to do with the plugin removing them during this period of communication between the master and slave. So what you'll see is a bunch of containers running in Mesos that you have to kill manually.

**Comment by maselvaraj** [comment]
@barkerd427  Thanks for the info!

**Comment by rasputnik** [comment]
Should we expect the jenkins.slaves.restarter.JnlpSlaveRestarterInstaller errors to not be there, or should   -noReconnect be the default?

**Comment by farhadkzm** [comment]
I fixed the issue by using `3.36-2` as the Jenkins slave tag [lts ](https://hub.docker.com/layers/jenkins/jenkins/lts/images/sha256-d5069c543e80454279caacd13457d012fb32c5229b5037a163d8bf61ffa6b80b)for the master
